FHEM/10_SD_Rojaflex.pm

```diff
@@ -0,0 +1,639 @@
+					readingsBulkUpdate($defs{$d}, 'motor', $motor, 1);
+					readingsBulkUpdate($defs{$d}, 'tpos', $tpos, 1);
+					if (AttrVal($defs{$d}{NAME},'bidirectional',1) eq '0') {
+						readingsBulkUpdate($defs{$d}, 'pct', $cpos, 1);
+						readingsBulkUpdate($defs{$d}, 'cpos', $cpos, 1);
+					}
+				}
+				readingsEndUpdate($defs{$d}, 1);
+			}
+		}
+	}
+	return;
+}
+
+sub SD_Rojaflex_pctStop {
+	my ($name) = @_;
+	my $hash = $defs{$name};
+	RemoveInternalTimer($name);
+	CommandSet($hash, "$name stop");
+	return;
+}
+
+sub SD_Rojaflex_clearfav {
+	my ($name) = @_;
+	my $hash = $defs{$name};
+	RemoveInternalTimer($name);
+	$hash->{clearfavcount} += 1;
+	if ($hash->{clearfavcount} < 4) { # 3 mal stop senden
+		CommandSet($hash, "$name stop");
+		InternalTimer( (gettimeofday() + 1), \&SD_Rojaflex_clearfav, $name );
+	} else {
+		CommandSet($hash, "$name savefav");
+		delete($hash->{clearfavcount});
+	}
+	return;
+}
+
+sub Define {
+	# define <name> SD_Rojaflex <hauscode>_<channel>
+	# define SD_Rojaflex_Test_11 SD_Rojaflex 7AE3121_11
+	# define <name> SD_Rojaflex <hauscode>_<channel> <iodevice>
+	# define SD_Rojaflex_Test_11 SD_Rojaflex 7AE3121_11 sduino434
+
+	my ($hash, $def) = @_;
+	my @a = split m{\s+}xms , $def;
+	my $name = $hash->{NAME};
+	my $iodevice;
+	my $ioname;
+
+	return 'Define SD_Rojaflex wrong syntax: define <name> SD_Rojaflex housecode_channel' if (int(@a) < 3);
+	my ($housecode, $channel) = split /[_]/xms , $a[2], 2;
+	return 'Define SD_Rojaflex wrong syntax, must be: housecode_channel' if (!defined $housecode || !defined $channel);
+	return 'Define SD_Rojaflex wrong housecode format: specify a 7 digit hex value [a-fA-F0-9]' if ($housecode !~ m/^[a-fA-F0-9]{7}$/xms );
+	return 'Define SD_Rojaflex wrong channel format: specify a decimal value [0-15]' if ($channel !~ m/^[0-9]{1,2}$/xms || $channel > 15);
+	if (scalar @a == 4) { $iodevice = $a[3] };
+
+	$hash->{DEF} = $a[2];
+	$hash->{VersionModule} = $VERSION;
+
+	$modules{SD_Rojaflex}{defptr}{$hash->{DEF}} = $hash;
+	if (exists $modules{SD_Rojaflex}{defptr}{ioname} && !defined $iodevice) { $ioname = $modules{SD_Rojaflex}{defptr}{ioname} };
+	if (!defined $iodevice) { $iodevice = $ioname }
+	AssignIoPort($hash, $iodevice);
+
+	if (not defined($attr{$name}{webCmd})) {$attr{$name}{webCmd} = 'up:stop:down'};
+
+	Log3 $name, 4, "SD_Rojaflex_Define: $a[0] HC=$housecode CHN=$channel";
+	return;
+}
+
+sub Undef {
+	my ($hash, $name) = @_;
+	if (defined($hash->{CODE}) && defined($modules{SD_Rojaflex}{defptr}{$hash->{CODE}})) { delete($modules{SD_Rojaflex}{defptr}{$hash->{CODE}}) };
+	RemoveInternalTimer($name);
+	return;
+}
+
+sub Parse {
+	my ($iohash, $msg) = @_;
+	my $ioname = $iohash->{NAME};
+	my ($protocol,$rawData) = split /[#]/xms , $msg;
+	$protocol =~ s/^[P](\d+)/$1/xms; # extract protocol
+	my $EMPTY = q{};
+
+	if (length ($rawData) < 18 ) { # 083122FD2C1A011AB1
+		Log3 $ioname, 1, "$ioname: SD_Rojaflex_Parse, rawData $rawData, message is to short";
+		return $EMPTY;
```

Comment:
Ich habe mir die Disptach Funktion angesehen, welche parseFn aufruft:

https://github.com/fhem/fhem-mirror/blob/e592647ce7edce1bb78839d32bc2fbc9ec9d5224/fhem/fhem.pl#L4091

Ich bin der Meinung, dass im Fehlerfall undef zurück geliefert werden sollte und nicht etwas anderes:

Daher schlage ich folgendes Vor:
```suggestion
		return ;
```

Reply:
Das gefällt mir eigentlich nicht so gut.
Bei meiner Variante bekomme ich im Fehlerfall nur diese Log-Ausgabe:

```
2021.10.30 17:23:33 1: sduino_dummy: SD_Rojaflex_Parse, rawData 083122FD290A010A8, message is to short
```
Bei deiner Variante erfolgt die Ausgabe in der Art:

```
2021.10.30 17:26:08 1: sduino_dummy: SD_Rojaflex_Parse, rawData 083122FD291A011AA, message is to short
2021.10.30 17:26:08 3: sduino_dummy: Unknown code P109#083122FD291A011AA, help me!
```

Reply:
Letzteres scheint mir aber richtig. Unter Umständen hätte FHEM ja versucht den Code an noch ein anderes Modul zu übergeben, nachdem SD_Rojaflex_Parse ein undef geliefert hat. Am Ende der Verarbeitung erscheint dann ein unknown code.


Wenn wir die Meldung SD_Rojaflex_Parse in Verbose 4 stecken, dann passt das doch im Endeffekt wieder alles.

Reply:
OK, dann machen wir das halt so. Verbose 1 belasse ich so, weil:

```
verbose
Setzt den Schwellwert für die Logfile-Meldungen.Mögliche Werte sind: •0 - Server start/stop
•1 - Fehlermeldungen oder unbekannte Pakete
•2 - bedeutende Ereigbisse/Alarme.
•3 - ausgesendete Kommandos werden gelogged.
•4 - von den einzelnen Geräten empfangene Daten.
•5 - Fehlersuche.

```